spec/formatting.md

```diff
@@ -18,3 +18,71 @@ the local variable takes precedence.
 
 It is an error for a local variable definition to
 refer to a local variable that's defined after it in the message.
+
+## Error Handling
+
+During the formatting of a message,
+various errors may be encountered.
+These are divided into the following categories:
+
+- **Syntax errors** occur when the syntax representation of a message is invalid.
+- **Resolution errors** occur when the runtime value of a part of a message
```

Comment:
The difference between the "resolution" and "formatting" errors that's proposed here is perhaps best seen by considering how a Placeholder containing an Expression is handled. Let's say that we have something like
```
{(foo) :message}
```
or
```
{(user_age) :global}
```
where `:message` and/or `:global` is a custom Function that uses its argument to look up a value from elsewhere, and that this value is then formatted as a part of the final message.

During the formatting of this, the custom code could then emit two different sorts of errors:
1. Resolution Error if there's a failure in getting the value that is to be formatted, e.g. if no `foo` message is available or if the `user_age` global is not set.
2. Formatting Error if the found value can't be formatted, e.g. because the `foo` message includes a variable reference that can't be resolved, or the `user_age` value turns out to have some unexpected shape.

Would you agree that these are different error categories, and that this categorical split could lead to different error handling in user code?

Reply:
Both `foo` and `user_age` are literal values here, so the errors are coming from the `:message` and `:global` custom functions; from the PoV of the core implementation, the are no variables here to resolve.

The intent here is to allow for a custom formatting function to emit two different kinds of errors: Resolution and Formatting. This is meant to enable something like `:message` or `:global` to work from an end-user PoV as much as possible like core features such as `$var`.

Reply:
Nitpick: I don't think what I am saying here affects the spec, but the answer above.

I think it is contradicting:
  * Resolution  ... failure in getting the value that is to be formatted
  * Formatting ... includes a variable reference that can't be resolved ...

It is unclear what is the difference between the two: "get the value" and "resolve variable reference"
The "publicly visible" operation is probably "I have a variable named `foo`, I want to get the value"
There is no "reference" except deep in the implementation.

But that implementation detail should not "leak" in the kind of error I am getting.

Reply:
Please note that the syntax example for the above discussion is this placeholder:
```
{(foo) :message}
```

The `foo` here is not a variable reference, it's a literal value that a custom `:message` function is interpreting as a message identifier.